go/src/net/ConnectionPool.go

```diff
@@ -0,0 +1,230 @@
+	if pool.conf == nil {
+		notInitialized := errors.New("Failed to get session: There is no config in the connection pool")
+		log.Println(notInitialized)
+		return nil, notInitialized
+	}
+
+	// Get valid and usable connection
+	conn, err := pool.GetIdleConn()
+	if err != nil {
+		log.Printf("Failed to get session, error: %s", err.Error())
+		return nil, err
+	}
+
+	// Authenticate
+	resp, err := conn.Authenticate(username, password)
+	if err != nil {
+		log.Printf("Failed to authenticate with the given credential, error: %s", err.Error())
+		return nil, err
+	}
+
+	sessionID := resp.GetSessionID()
+	// Create new session
+	newSession := newSession(sessionID, conn, pool)
+	// Mark connection as in use
+	pool.ConnectionInUse[conn] = true
+	// Add connction to active queue and pop it from idle queue
+	pool.activeConnectionQueue.PushBack(conn)
+
+	pool.ConnectionInUse[conn] = true
+	// Update host status
+	pool.loadBalancer.ValidateHost(&conn.SeverAddress)
+
+	for _, status := range pool.loadBalancer.ServerStatusList {
+		if *status.address == conn.SeverAddress {
+			status.workLoad = status.workLoad + 1
+		}
+	}
+	return &newSession, nil
+
+}
+
+func (pool *ConnectionPool) GetIdleConn() (*Connection, error) {
+	// Update status of all server's avaliability
+	pool.loadBalancer.UpdateServerStatus()
+
+	// Take an idle connection if possible
+	if pool.idleConnectionQueue.Len() > 0 {
+		pool.UpdateConnectionStatus()
+		newConn, err := pool.GetValidConn()
+		if err != nil {
+			log.Printf("Error: %s \n", err.Error())
+			return nil, err
+		}
+		for ele := pool.idleConnectionQueue.Front(); ele != nil; ele = ele.Next() {
+			if ele.Value.(*Connection) == newConn {
+				pool.idleConnectionQueue.Remove(ele)
+			}
+		}
+		return newConn, nil
+	}
+
+	// Create a new connection if there is no idle connection and total connection < pool max size
+	totalConn := pool.idleConnectionQueue.Len() + pool.activeConnectionQueue.Len()
+	if totalConn < pool.conf.MaxConnPoolSize {
+		// Get a valid host
+		host, err := pool.loadBalancer.GetValidHost()
+		// No valid host in the pool
+		if err != nil {
+			log.Printf("Error: %s \n", err.Error())
+			return nil, err
+		}
+		newConn := NewConnection(*host)
+
+		// Open connection to host
+		err = newConn.Open(newConn.SeverAddress, *pool.conf)
+		if err != nil {
+			log.Printf("Failed to open connection, error: %s \n", err.Error())
+			return nil, err
+		}
+		return newConn, nil
+	}
+	// TODO: If no idle avaliable, wait for timeout and reconnect
+
+	// If no idle avaliable and the number of total connection reaches the max pool size, return error/wait for timeout
+	noAvaliableConnectionErr := errors.New("Failed to get connection: no avaliable connection in the connection pool")
+	log.Print(noAvaliableConnectionErr)
+	return nil, noAvaliableConnectionErr
+}
+
+// Release connection to pool
+func (pool *ConnectionPool) ReturnObject(conn *Connection) {
+	pool.idleConnectionQueue.PushBack(conn)
```

Comment:
need to judge the connection is an active connection firstly, if not, do nothing.